### PR TITLE
Auth: Refactoring of frontend skipOrgRoleSync logic to backend

### PR DESCRIPTION
**Pull request by eleijonmarck**
**What is this feature?**
Refactoring of frontend logic to backend logic.

- adds tests for extracting the externalSyncedUser
- remove logic from frontend
- adds const for labels and providers

Fixes #
https://github.com/grafana/grafana-enterprise/issues/4235

Notes:

> Do you know how cases, where a user is using several auth providers, would be handled? In those cases, we should only return false if none of them is used to sync org roles. Is that the way it would currently work?

After investigation, the userAuth is overwritten each time the user logs in with a auth provider, as such whichever auth provider the user chose to login with becomes the auth provider we consider to be the "truth", and this is shown in the UI

**Special notes for your reviewer**:

Sorry for the long video, but it showcases both when adding specific providers and changing them and the overall setting and that it works for two providers separately 🙏 
demo - 

https://user-images.githubusercontent.com/7727602/217038885-742d174b-e018-41be-87d4-b36a75d04c16.mp4

### CODE REVIEW

File: pkg/api/frontendsettings.go [hunk 1]
```diff
@@ -148,7 +148,7 @@ func (hs *HTTPServer) getFrontendSettings(c *contextmodel.ReqContext) (*dtos.Fro
 
 		Auth: dtos.FrontendSettingsAuthDTO{
 			OAuthSkipOrgRoleUpdateSync:  hs.Cfg.OAuthSkipOrgRoleUpdateSync,
-			SAMLSkipOrgRoleSync:         hs.Cfg.SectionWithEnvOverrides("auth.saml").Key("skip_org_role_sync").MustBool(false),
+			SAMLSkipOrgRoleSync:         hs.Cfg.SAMLSkipOrgRoleSync,
```
Comment (eleijonmarck) [inline pkg/api/frontendsettings.go:151]: added missing configuration field for config for SAML

File: pkg/services/login/authinfo_test.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
+package login
+
+import (
+	"testing"
+
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestIsExternallySynced(t *testing.T) {
```
Comment (eleijonmarck) [inline pkg/services/login/authinfo_test.go:10]: I made an exhaustive list of tests. could be made smaller and only test one/two providers

File: pkg/services/user/model.go [hunk 3]
```diff
@@ -140,20 +140,21 @@ type GetUserProfileQuery struct {
...
+	ID                 int64           `json:"id"`
+	Email              string          `json:"email"`
+	Name               string          `json:"name"`
+	Login              string          `json:"login"`
+	Theme              string          `json:"theme"`
+	OrgID              int64           `json:"orgId,omitempty"`
+	IsGrafanaAdmin     bool            `json:"isGrafanaAdmin"`
+	IsDisabled         bool            `json:"isDisabled"`
+	IsExternal         bool            `json:"isExternal"`
+	IsExternallySynced bool            `json:"isExternallySynced"`
```
Comment (eleijonmarck) [inline pkg/services/user/model.go:152]: this field is the one added

File: public/app/features/admin/UserAdminPage.tsx [hunk 4]
```diff
@@ -103,46 +103,9 @@ export class UserAdminPage extends PureComponent<Props> {
...
-        (!config.auth.JWTAuthSkipOrgRoleSync && isJWTUser) ||
-        // both OAuthSkipOrgRoleUpdateSync and specific provider settings needs to be false for a user to be synced
-        (!config.auth.OAuthSkipOrgRoleUpdateSync && !config.auth.GrafanaComSkipOrgRoleSync && isGrafanaComUser) ||
-        (!config.auth.OAuthSkipOrgRoleUpdateSync && !config.auth.OktaSkipOrgRoleSync && isOktaUser) ||
-        (!config.auth.OAuthSkipOrgRoleUpdateSync && !config.auth.GithubSkipOrgRoleSync && isGithubUser) ||
-        (!config.auth.OAuthSkipOrgRoleUpdateSync && !config.auth.AzureADSkipOrgRoleSync && isAzureADUser) ||
-        (!config.auth.OAuthSkipOrgRoleUpdateSync && !config.auth.GitLabSkipOrgRoleSync && isGitLabUser) ||
-        (!config.auth.OAuthSkipOrgRoleUpdateSync && !config.auth.GenericOAuthSkipOrgRoleSync && isGenericOAuthUser) ||
-        (!config.auth.OAuthSkipOrgRoleUpdateSync && !config.auth.GoogleSkipOrgRoleSync && isGoogleUser));
+    const isUserSynced = !config.auth.DisableSyncLock && user?.isExternallySynced;
```
Comment (eleijonmarck) [inline public/app/features/admin/UserAdminPage.tsx:108]: this is the 🍖  of this PR

**Review by IevaVasiljeva — COMMENTED** [review]
Nice, doing the logic on the backend is way neater imo ✨ 

I've left some comments and suggestions. Do you know how cases where a user is using several auth providers would be handled? I think in those cases we should only return false if none of them is used to sync org roles. Is that the way it would currently work?

File: public/app/features/admin/UserAdminPage.tsx [hunk 5]
```diff
@@ -164,13 +127,9 @@ export class UserAdminPage extends PureComponent<Props> {
...
                 onPasswordChange={this.onPasswordChange}
               />
-              {!config.auth.LDAPSkipOrgRoleSync &&
-                isLDAPUser &&
-                featureEnabled('ldapsync') &&
-                ldapSyncInfo &&
-                canReadLDAPStatus && (
-                  <UserLdapSyncInfo ldapSyncInfo={ldapSyncInfo} user={user} onUserSync={this.onUserSync} />
-                )}
+              {isUserSynced && featureEnabled('ldapsync') && ldapSyncInfo && canReadLDAPStatus && (
```
[the remark this answers is not in the document]
Comment (eleijonmarck) [reply public/app/features/admin/UserAdminPage.tsx:130]: oh i did leave that in first. I think you are right here, we need to check if the user is a LDAP user as well.
Nice catch ⚡

File: pkg/services/login/authinfo.go [hunk 6]
```diff
@@ -22,32 +23,92 @@ const (
...
+)
+
+// User is allowed to change org role in the UI
+// by not being synced with an external system
+func IsExternallySynced(cfg *setting.Cfg, autoProviderLabel string) bool {
+	// FIXME: remove this once we remove the setting
+	// is a deprecated setting that is used to skip org role sync for all external oauth providers
+	if cfg.OAuthSkipOrgRoleUpdateSync {
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (eleijonmarck) [reply pkg/services/login/authinfo.go:59]: yes! you are right, I will change so that we check SAML, LDAP and JWT first and then check the OAuth providers

**Comment by eleijonmarck** [comment]
> Do you know how cases where a user is using several auth providers would be handled? I think in those cases we should only return false if none of them is used to sync org roles. Is that the way it would currently work?

this is a very good question. My assumptions is that it currently acts based on the [GetAuthProvider](https://github.com/grafana/grafana/pull/62921/files#diff-b2aed1a2057692b017074d5d8d78c518d9bb4756e553b1473cce43f311a63660R89) function. So if a user has multiple oauth providers linked it depends on the order of the switch statement as to which provider the user gets. Does this make sense?

**Comment by eleijonmarck** [comment]
> I think that it currently acts based on the [GetAuthProvider](https://github.com/grafana/grafana/pull/62921/files#diff-b2aed1a2057692b017074d5d8d78c518d9bb4756e553b1473cce43f311a63660R89) function. So if a user has multiple oauth providers linked it depends on the order of the switch statement as to which provider the user gets. Do you think this makes sense?
this is incorrect.

> Do you know how cases where a user is using several auth providers would be handled? I think in those cases we should only return false if none of them is used to sync org roles. Is that the way it would currently work?

The userAuth is overwritten each time the user logs in with a auth provider, as such whichever auth provider the user chose to login with becomes the auth provider we consider to be the "truth", and this is shown in the UI

File: pkg/services/login/authinfo.go [hunk 7]
```diff
@@ -17,37 +18,102 @@ type AuthInfoService interface {
...
+	GrafanaComLabel   = "grafana.com"
+	OktaLabel         = "Okta"
 )
 
+// IsExternnalySynced is used to tell if the user is externallySynced or that we should the external sync
+// true means that the org role sync is handled by Grafana
+// Note: currently the users authinfo is overridden each time the user logs in
+// https://github.com/grafana/grafana/blob/4181acec72f76df7ad02badce13769bae4a1f840/pkg/services/login/authinfoservice/database/database.go#L61
+// this means that if the user has multiple auth providers and one of them is set to sync org roles
+// then the user will be synced from the provider they last logged in with
```
Comment (IevaVasiljeva) [inline pkg/services/login/authinfo.go:57]: nit: I think if only one provider is set to sync org roles, roles will only be synced when you log in with this provider. This comment makes it sound like they will be synced for all providers. Maybe something like:
```suggestion
// this means that if the user has multiple auth providers and one of them is set to sync org roles
// then IsExternallySynced will be true for this one provider and false for the others
```

**Review by IevaVasiljeva — APPROVED** [review]
Nice, mostly LGTM! One small question about GCom provider and some suggestions for the phrasing.

File: pkg/services/login/authinfo.go (hunk 7 above)
Comment (eleijonmarck) [reply pkg/services/login/authinfo.go:57]: i see. let me look at it again and your suggestion

Comment (eleijonmarck) [reply pkg/services/login/authinfo.go:57]: yea this is more local to the function and is what happens. nice ! 👍

File: pkg/services/login/authinfo.go [hunk 8]
```diff
@@ -17,37 +18,102 @@ type AuthInfoService interface {
...
+	case OktaLabel:
+		return !cfg.OktaSkipOrgRoleSync
+	case AzureADLabel:
+		return !cfg.AzureADSkipOrgRoleSync
+	case GitLabLabel:
+		return !cfg.GitLabSkipOrgRoleSync
+	case GithubLabel:
+		return !cfg.GithubSkipOrgRoleSync
+	case GrafanaComLabel:
+		return !cfg.GrafanaComSkipOrgRoleSync
```
[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/login/authinfo.go:86]: Gotcha, learn something new every day 😄

**Pull request merged**